keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlResourceRepository.kt

```diff
@@ -288,29 +279,33 @@ open class SqlResourceRepository(
   }
 
   override fun delete(id: String) {
-    val uid = sqlRetry.withRetry(READ) {
-      jooq.select(RESOURCE.UID)
-        .from(RESOURCE)
-        .where(RESOURCE.ID.eq(id))
-        .fetchOne(RESOURCE.UID)
-        ?.let(ULID::parseULID)
-        ?: throw NoSuchResourceId(id)
-    }
-    sqlRetry.withRetry(WRITE) {
-      jooq.deleteFrom(RESOURCE)
-        .where(RESOURCE.UID.eq(uid.toString()))
-        .execute()
-    }
-    sqlRetry.withRetry(WRITE) {
-      jooq.deleteFrom(EVENT)
-        .where(EVENT.SCOPE.eq(Scope.RESOURCE.name))
-        .and(EVENT.UID.eq(uid.toString()))
-        .execute()
-    }
-    sqlRetry.withRetry(WRITE) {
-      jooq.deleteFrom(DIFF_FINGERPRINT)
-        .where(DIFF_FINGERPRINT.RESOURCE_ID.eq(id))
-        .execute()
+    jooq.transaction { config ->
+      val txn = DSL.using(config)
+
+      val uid = sqlRetry.withRetry(READ) {
+        txn.select(RESOURCE.UID)
+          .from(RESOURCE)
+          .where(RESOURCE.ID.eq(id))
+          .fetchOne(RESOURCE.UID)
+          ?.let(ULID::parseULID)
+          ?: throw NoSuchResourceId(id)
+      }
+      sqlRetry.withRetry(WRITE) {
+        txn.deleteFrom(RESOURCE)
+          .where(RESOURCE.UID.eq(uid.toString()))
+          .execute()
+      }
+      sqlRetry.withRetry(WRITE) {
+        txn.deleteFrom(EVENT)
+          .where(EVENT.SCOPE.eq(Scope.RESOURCE.name))
+          .and(EVENT.UID.eq(uid.toString()))
```

Comment:
Deleting a row by anything other than the primary or a secondary unique key locks the table to inserts. When the delete is part of a transaction as is now the case, the insert lock is for the duration of the transaction. Periodically locking the event table for 200ms seems subpar for keel's scalability, especially as spring event processing (the source of many event table inserts) is serialized and blocking. Also to be fixed, the events table doesn't have a declared PK or unique index, I assume the uid column was intended to be a PK.

Reply:
Or seems more like the `uid` column on event should rightly be called `resource_id`. Event's should probably get their own uid per event. You could then optionally drop the timestamp column, as ulids are also time ordered millisecond accurate timestamps.

Reply:
> When the delete is part of a transaction as is now the case, the insert lock is for the duration of the transaction. Periodically locking the event table for 200ms seems subpar for keel's scalability, especially as spring event processing (the source of many event table inserts) is serialized and blocking.

So would you rather have partial deletes in the logical set not be rolled back at all if one fails? Generally there will be a trade off between performance and consistency, and my feeling was that the latter was more important in this case. Happy to remove the transaction  if we don't think that's the case, though. This was tangential to my PR anyway.

> Also to be fixed, the events table doesn't have a declared PK or unique index, I assume the uid column was intended to be a PK?

No, because as you mentioned it was originally meant to be an FK to the resource UID. Since we don't actually have FKs anywhere, I reused this column for application-level events, where it will contain the application name. If we want to model things with multiple tables (one per event type), that's fine -- then we could add better constraints, but I wasn't sure it was worth the trouble since we're not enforcing integrity via FKs anyway.

Reply:
> So would you rather have partial deletes in the logical set not be rolled back at all if one fails?

If my comment read as a criticism of the use of a transaction here, that was not intentional. Deletes of the events table should be made non-locking regardless. The lack of a unique/primary key on this table for the sake of deletes will further bite keel if a limited event retention policy has to be implemented.

Reply:
No worries. My response didn't come out the way I intended either. 😃 
I've removed the transaction around the deletes for now as it warrants a separate PR with a focused discussion. Thanks for the feedback!